datafusion/expr/src/expr.rs

```diff
@@ -1030,6 +1034,49 @@ impl Expr {
     pub fn contains_outer(&self) -> bool {
         !find_out_reference_exprs(self).is_empty()
     }
+
+    /// Find all [`Expr::Placeholder`] in anthis, and try
+    /// to infer their [`DataType`] from the context of their use.
+    pub fn infer_placeholder_types(self, schema: &DFSchema) -> Result<Expr> {
```

Comment:
This was moved from the `datafusion-sql` module as it is not SQL specific, but generic to expressions